shared/src/commonMain/kotlin/org/hyperskill/app/track_list/presentation/TrackListActionDispatcher.kt

```diff
@@ -0,0 +1,103 @@
+                    HyperskillSentryTransactionBuilder.buildPlaceholderNewUserScreenRemoteDataLoading()
+                sentryInteractor.startTransaction(sentryTransaction)
+
+                val tracks = trackInteractor
+                    .getAllTracks()
+                    .getOrElse {
+                        sentryInteractor.finishTransaction(sentryTransaction, it)
+                        return onNewMessage(Message.TracksLoaded.Error)
+                    }
+                    .filter { it.isBeta.not() }
+
+                val trackProgressById = progressesInteractor
+                    .getTracksProgresses(tracks.map { it.id }, forceLoadFromRemote = true)
+                    .getOrElse {
+                        sentryInteractor.finishTransaction(sentryTransaction, it)
+                        return onNewMessage(Message.TracksLoaded.Error)
+                    }
+                    .associateBy { it.id }
+
+                if (tracks.size != trackProgressById.size) {
+                    sentryInteractor.captureErrorMessage("TrackList: tracks.size != tracksProgresses.size")
+                }
+
+                val tracksWithProgresses = tracks.mapNotNull { track ->
+                    trackProgressById[track.progressId]?.let { trackProgress ->
+                        TrackWithProgress(
+                            track = track,
+                            trackProgress = trackProgress
+                        )
+                    }
+                }
+
+                val currentTrackId = currentStudyPlanStateRepository.getState()
+                    .getOrElse {
+                        sentryInteractor.finishTransaction(sentryTransaction, it)
+                        return onNewMessage(Message.TracksLoaded.Error)
+                    }
+                    .trackId
+
+                sentryInteractor.finishTransaction(sentryTransaction)
+
+                onNewMessage(Message.TracksLoaded.Success(tracksWithProgresses, currentTrackId))
+            }
+            is Action.SelectTrack -> {
+                val currentProfile = profileInteractor
+                    .getCurrentProfile()
+                    .getOrElse {
+                        return onNewMessage(Message.TrackSelected.Error)
+                    }
+
+                val isFreemiumEnabled = freemiumInteractor
+                    .isFreemiumEnabled()
+                    .getOrElse {
+                        return onNewMessage(Message.TrackSelected.Error)
+                    }
+
+                profileInteractor.selectTrack(currentProfile.id, action.track.id)
+                    .getOrElse {
+                        return onNewMessage(Message.TrackSelected.Error)
+                    }
+
+                onNewMessage(Message.TrackSelected.Success)
+
+                if (!isFreemiumEnabled && action.track.projects.isNotEmpty()) {
+                    onNewMessage(Message.ProjectSelectionRequired(action.track.id))
+                }
```

Comment:
Maybe we should do this check before `profileInteractor.selectTrack`?

Reply:
In fact, this check isn't relevant because of we will navigate user to the study plan screen anyway, I have removed it